mars/_utils.pyx

```diff
@@ -93,57 +93,50 @@ cdef inline to_hex(bytes s):
         return hexlify(s)
 
 
-cdef inline object h(object ob):
-    if isinstance(ob, dict):
-        return h_iterative(sorted(list(ob.items()), key=str))
-    elif isinstance(ob, set):
-        return h_iterative(sorted(ob, key=str))
-    elif isinstance(ob, (tuple, list)):
-        return h_iterative(ob)
-    return h_non_iterative(ob)
+cdef class Tokenizer:
+    cdef dict _handlers
+    def __init__(self):
+        self._handlers = dict()
 
+    def register(self, cls, handler):
+        self._handlers[cls] = handler
 
-cdef inline list h_iterative(object ob):
+    cdef inline tokenize(self, object obj):
+        object_type = type(obj)
+        try:
+            handler = self._handlers[object_type]
+            return handler(obj)
+        except KeyError:
+            if hasattr(obj, '_key'):
+                return getattr(obj, '_key')
```

Comment:
Just return `obj._key`.

Reply:
Just to avoid access to a protected member, will fix